src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+                if (len != -1)
+                {
+                    await buf.Ensure(len, true, cancellationToken);
+                    buf.Skip(len); // We know already what we created
+                }
+
+                Expect<CommandCompleteMessage>(await connector.ReadMessage(true, cancellationToken), connector);
+                Expect<ReadyForQueryMessage>(await connector.ReadMessage(true, cancellationToken), connector);
+
+                return new NpgsqlReplicationSlotOptions(slotNameResult, consistentPoint, snapshotName);
+            }
+            catch (PostgresException e)
+            {
+                if (PostgreSqlVersion < FirstVersionWithTemporarySlots && e.SqlState == PostgresErrorCodes.SyntaxError)
+                {
+                    if (temporarySlot)
+                        throw new ArgumentException("Temporary replication slots were introduced in PostgreSQL " +
+                                                    $"{FirstVersionWithTemporarySlots.ToString(1)}. " +
+                                                    $"Using PostgreSQL version {PostgreSqlVersion.ToString(3)} you " +
+                                                    $"have to set the {nameof(temporarySlot)} argument to false.",
+                            nameof(temporarySlot), e);
+                }
+                throw;
+            }
+        }
+
+        internal async IAsyncEnumerable<NpgsqlXLogDataMessage> StartReplicationInternal(
+            Action<StringBuilder> startCommandAction,
+            bool bypassingStream,
+            [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var stateResetter = EnsureAndSetState(ReplicationConnectionState.Idle, ReplicationConnectionState.Streaming);
+            var connector = _npgsqlConnection.Connector!;
+#if !NETSTANDARD2_0
+            await
+#endif
+            using var registration = cancellationToken.Register(c => ((NpgsqlReplicationConnection)c!).Cancel(), this);
+            var createCommandBuilder = new StringBuilder("START_REPLICATION ");
+            startCommandAction(createCommandBuilder);
+            await connector.WriteQuery(createCommandBuilder.ToString(), true, cancellationToken);
+            await connector.Flush(true, cancellationToken);
+
+            var msg = await connector.ReadMessage(true, cancellationToken);
+            switch (msg.Code)
+            {
+            case BackendMessageCode.CopyBothResponse:
+                break;
+            case BackendMessageCode.CommandComplete:
+            {
+                yield break;
+            }
+            default:
+                stateResetter.ChangeResetState(ReplicationConnectionState.Broken);
+                throw connector.UnexpectedMessageReceived(msg.Code);
+            }
+
+            _sendFeedbackTimer.Change(WalReceiverStatusInterval, Timeout.InfiniteTimeSpan);
```

Comment:
We need to reset these when completing replication, in a finally, no? We could dispose them to wait for their callbacks to complete as well.

Reply:
Yeah, I just kept them running and return in the callback immediately, which probably didn't hurt but is admittedly a bad style.

Reply:
https://github.com/Brar/npgsql/pull/10